presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterCollector.java

```diff
@@ -0,0 +1,261 @@
+                        rowExpression.get(),
+                        staticFilter.get());
+                resultExpression = Optional.of(combinedExpression);
+                globalState = DynamicProcessorState.DONE;
+                triggerCallback(combinedExpression);
+            }
+            catch (Exception e) {
+                globalState = DynamicProcessorState.FAILED;
+                throw e;
+            }
+        }
+        else {
+            log.debug("DF Processing: Tuple Domain is all");
+            globalState = DynamicProcessorState.FAILED;
+        }
+    }
+
+    public Optional<RowExpression> getFinalExpression()
+    {
+        if (globalState != DynamicProcessorState.DONE) {
+            return Optional.empty();
+        }
+        else {
+            return resultExpression;
+        }
+    }
+
+    private TupleDomain translateSummaryIntoTupleDomain(DynamicFilterSummary summary, Set<DynamicFilterExpression> dynamicFilters)
+    {
+        if (!summary.getTupleDomain().getDomains().isPresent()) {
+            return TupleDomain.all();
+        }
+
+        Map<String, Symbol> sourceExpressionSymbols = extractSourceExpressionSymbols(dynamicFilters);
+        ImmutableMap.Builder<Symbol, Domain> domainBuilder = ImmutableMap.builder();
+        for (Map.Entry<String, Domain> entry : summary.getTupleDomain().getDomains().get().entrySet()) {
+            Symbol actualSymbol = sourceExpressionSymbols.get(entry.getKey());
+            if (actualSymbol == null) {
+                continue;
+            }
+            domainBuilder.put(actualSymbol, entry.getValue());
+        }
+
+        final ImmutableMap<Symbol, Domain> domainMap = domainBuilder.build();
+        return domainMap.isEmpty() ? TupleDomain.all() : TupleDomain.withColumnDomains(domainMap);
+    }
+
+    private Map<String, Symbol> extractSourceExpressionSymbols(Set<DynamicFilterExpression> dynamicFilters)
+    {
+        ImmutableMap.Builder<String, Symbol> resultBuilder = ImmutableMap.builder();
+        for (DynamicFilterExpression dynamicFilter : dynamicFilters) {
+            Expression expression = dynamicFilter.getProbeExpression();
+            if (!(expression instanceof SymbolReference)) {
+                continue;
+            }
+            resultBuilder.put(dynamicFilter.getDfSymbol(), Symbol.from(expression));
+        }
+        return resultBuilder.build();
+    }
+
+    public void addListener(DFCollectorCallback callback)
+    {
+        listeners.add(callback);
+    }
+
+    private void triggerCallback(RowExpression result)
+    {
+        listeners.forEach(callback -> callback.onSuccess(result));
+    }
+
+    public boolean isFailed()
+    {
+        return globalState == DynamicProcessorState.FAILED;
+    }
+
+    public boolean isDone()
+    {
+        return globalState == DynamicProcessorState.DONE;
+    }
+
+    public interface DFCollectorCallback
+    {
+        void onSuccess(RowExpression result);
```

Comment:
Would it be possible to model collector as a service with method:
```
ListenableFuture<RowExpression> collectDFFilter(DynamicFilter dynamicFilter)
```
It would be more consistent with how asynchronous events are modeled in Presto. Such future could be cancelled when query waiting for it finishes early.